### PR TITLE
is-instance check to determine the type of a plugin for teardown decision

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Addresses a comment on recently merged PR https://github.com/PyTorchLightning/pytorch-lightning/pull/8685#discussion_r682152037 which got missed/misinterpreted regarding the check for a type of plugin. 

The plugin types classified as "interactive_compatible_types" do not relate directly to the plugin types that require a special place for teardown. These are the plugins that use the multiprocessing "spawn" method to create new processes but the DP plugin is NOT of this type and with the recent changes now does not have a teardown anymore. 
Also, with #8685 custom accelerators would not receive a teardown. 

This PR fixes that.

NOTE: labelling this as 1.5 bugfix as #8685  was also labelled this way.

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/trainer.py [hunk 1]
```diff
@@ -947,7 +946,7 @@ def _run(self, model: "pl.LightningModule") -> Optional[Union[_EVALUATE_OUTPUT,
 
         # teardown if necessary (similar calls for spawn plugins are excluded as they have
         # been included at the end of `new_process` functions)
-        if self._distrib_type not in DistributedType.interactive_compatible_types():
+        if not isinstance(self.training_type_plugin, DDPSpawnPlugin):
```
Comment (ananthsub) [inline pytorch_lightning/trainer/trainer.py:949]: - should we add a test to confirm DDP spawn and other parallel plugins all work as expected to avoid potential regressions here in the future?
- for custom plugins, this doesn't generalize unless they all extend DDP Spawn. should the plugin denote that it uses multiprocessing spawn method to create new processes, and that we need this in order to determine the appropriate teardown?

**Review by ananthsub — APPROVED** [review]

File: pytorch_lightning/trainer/trainer.py (hunk 1 above)
Comment (awaelchli) [reply pytorch_lightning/trainer/trainer.py:949]: yes it does not generalize to plugins that do spawn but are not subclassing our ddp spawn plugin, but this was a tradeoff introduced in #8685. However, the change in this PR covers the case of plugins that don't spawn and therefore get a teardown at the right place.

Comment (ananthsub) [reply pytorch_lightning/trainer/trainer.py:949]: Definitely, agreed that we should fix this first. I meant the custom plug-in support as a follow up beyond the scope of this PR

**Review by carmocca — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGTM !
